### PR TITLE
Lfu

### LINKED ISSUE #6: Least frequently used
Could you please implement 'least frequently used'  cache replacement algorithm?

### PR DESCRIPTION
**Pull request by benhurdavies**
# What's it about?
### Implementing LFU engine.
- [x] extend the DoubleLinkedList with addNext and addPrev methods.
- [x] Implement LFU.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -31,17 +31,55 @@ const cache = createCache({
...
 
 To implement custom hashTable you have to use methods with symbols name provided from the package. [example](src/featureTest/customHashTable.test.js)
 
-
 #### Next?
-* TTL engine.
-* TTL combining with LRU engine
-* LFU (Least frequently used) engine.
+
+- TTL engine.
```
Comment (TarSzator) [inline README.md]: Please remove the dot. Yeah this is how detail oriented I am ;-)

File: src/engines/LeastFrequentlyUsed.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  get as hGet,
+  has as hHas,
+  remove as hRemove,
+} from '../hashTableSymbol';
+
+function LeastFrequentlyUsed({
+  size = Number.MAX_SAFE_INTEGER,
+  HashTable = DefaultHashTable,
+}) {
+  if (size <= 0) throw Error('LRU size specified should be greater than zero');
```
Comment (TarSzator) [inline src/engines/LeastFrequentlyUsed.js]: null makes it true and is not replaced by default value ;-)
Sugestion validator function 'isPositiveInteger'

File: src/engines/LeastFrequentlyUsed.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  function removeFromFrequency(fItemNode) {
+    const { parent } = fItemNode.value;
+    const { items } = parent.value;
+    items.remove(fItemNode);
+    if (items.size() === 0) {
+      frequencies.remove(parent);
+    }
+  }
+
+  function newFrequencyItem({ payload, parent }) {
```
Comment (TarSzator) [inline src/engines/LeastFrequentlyUsed.js:112]: If you do not plan to add something here this call is kind of redundant 😄

File: src/engines/LeastFrequentlyUsed.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    const newItem = newFrequencyItem({ payload, parent: first });
+    return first.value.items.addFirst(newItem);
+  }
+
+  function incrementFrequency(fItemNode) {
+    const currentFrequencyNode = fItemNode.value.parent;
+    let nextFrequencyNode = currentFrequencyNode.next;
+    const nextFrequency = currentFrequencyNode.value.value + 1;
```
Comment (TarSzator) [inline src/engines/LeastFrequentlyUsed.js:71]: Crazy theory: nextFrequency could be higher then Number.MAX_SAFE_INTEGER.
I guess we should check for that and if though keep it in the currentFrequencyNode!?

**Review by TarSzator — CHANGES_REQUESTED** [review]
Cool stuff. Tough to read due to all the value.payload.value kind of stuff. but in this case. I guess sacrificing readability on the altar of speed is fine for this kind of library

File: src/engines/LeastFrequentlyUsed.js (hunk 3 above)
Comment (benhurdavies) [reply src/engines/LeastFrequentlyUsed.js:112]: `newFrequencyItem` are used in multiple places. By calling the same method for creating new object will be consistent across the code and more readable in my opinion

**Review by TarSzator — APPROVED** [review]
Nice

**Pull request merged**